### PR TITLE
Initial implementation of `redis-service.server.ts`

### PR DESCRIPTION
**Pull request by gregory-j-baker**
This is the initial implementation of a Redis client module. Currently it supports connecting and `get`/`set`/`del` operations.

Future enhancements should include connection pool configuration.

## Related ADO workitems:
- [AB#2673](https://dev.azure.com/DTS-STN/1fc40a8f-28cf-47bc-b6e4-1c234bd06177/_workitems/edit/2673)

### CODE REVIEW

File: frontend/app/services/redis-service.server.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  password?: string;
+}
+
+export function getRedisService(config: RedisServiceConfigurationOptions) {
+  return {
+    /**
+     * @see https://redis.io/commands/get/
+     */
+    get: async (key: string) => {
+      redisClient ??= await createRedisClient(config.url, config.username, config.password);
```
Comment (sebastien-comeau) [inline frontend/app/services/redis-service.server.ts:42]: Create this inside `getRedisService` and reuse it.

```ts
async function getRedisClient() {
 return redisClient ??= await createRedisClient(config.url, config.username, config.password);
}
```

File: frontend/app/services/redis-service.server.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      const value = await redisClient.get(key);
+      return value ? JSON.parse(value) : null;
+    },
+    /**
+     * @see https://redis.io/commands/set/
+     */
+    set: async (key: string, value: unknown, options?: SetOptions) => {
+      redisClient ??= await createRedisClient(config.url, config.username, config.password);
+      return await redisClient.set(key, JSON.stringify(value), options);
```
Comment (sebastien-comeau) [inline frontend/app/services/redis-service.server.ts:51]: You don't need to `await` before a return.

**Review by sebastien-comeau — CHANGES_REQUESTED** [review]

File: frontend/app/services/redis-service.server.ts (hunk 1 above)
Comment (gregory-j-baker) [reply frontend/app/services/redis-service.server.ts]: Fixed in latest push

File: frontend/app/services/redis-service.server.ts (hunk 2 above)
Comment (gregory-j-baker) [reply frontend/app/services/redis-service.server.ts]: Was left over from a refactoring. I removed it in my latest push.

File: frontend/app/services/redis-service.server.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+import { getEnv } from '~/utils/env.server';
+import { getLogger } from '~/utils/logging.server';
+
+const log = getLogger('redis-service.server');
+
+/**
+ * Singleton redis client instance; lazy initialized as needed.
+ */
+let redisClient: ReturnType<typeof createClient>;
```
Comment (sebastien-comeau) [inline frontend/app/services/redis-service.server.ts:29]:
```suggestion
let redisClient: ReturnType<typeof createClient> | undefined = undefined;
```

**Review by sebastien-comeau — CHANGES_REQUESTED** [review]

File: frontend/app/services/redis-service.server.ts (hunk 3 above)
Comment (gregory-j-baker) [reply frontend/app/services/redis-service.server.ts]: Fixed in latest push.

**Review by sebastien-comeau — APPROVED** [review]

**Pull request merged**